tracer/test/Datadog.Trace.Security.IntegrationTests/AspNetBase.cs

```diff
@@ -117,19 +117,72 @@ public async Task TestBlockedRequestAsync(MockTracerAgent agent, string url, int
 
                 return target.Tags;
             }));
-
             // Overriding the type name here as we have multiple test classes in the file
             // Ensures that we get nice file nesting in Solution Explorer
             await Verifier.Verify(spans, settings)
                           .UseMethodName("_")
                           .UseTypeName(GetTestName());
         }
 
-        protected void SetHttpPort(int httpPort)
+        protected async Task<System.Collections.Immutable.IImmutableList<MockSpan>> SendRequestsAsync(MockTracerAgent agent, string url, int expectedSpans, int attackNumber, bool parallel = false)
         {
-            _httpPort = httpPort;
+            var minDateTime = DateTime.UtcNow; // when ran sequentially, we get the spans from the previous tests!
+            var attacks = new ConcurrentBag<Task<(HttpStatusCode, string)>>();
+            if (parallel)
+            {
+                Parallel.For(0, attackNumber, _ => attacks.Add(SubmitRequest(url)));
+            }
+            else
+            {
+                for (int i = 0; i < attackNumber; i++)
+                {
+                    attacks.Add(SubmitRequest(url));
+                }
+            }
+
+            var resultRequests = await Task.WhenAll(attacks);
+            agent.SpanFilters.Add(s => s.Tags.ContainsKey("http.url") && s.Tags["http.url"].IndexOf("Health", StringComparison.InvariantCultureIgnoreCase) > 0);
+            var spans = agent.WaitForSpans(expectedSpans, minDateTime: minDateTime);
+            Assert.Equal(expectedSpans, spans.Count);
+            return spans;
+        }
+
+        protected async Task TestRateLimiter(bool enableSecurity, string url, MockTracerAgent agent, int traceRateLimit, int totalRequests, int totalExpectedSpans, bool parallel = true)
+        {
+            var excess = Math.Abs(totalRequests - traceRateLimit);
+            var errorMargin = 0.25;
```

Comment:
I know this may be flaky due to avoiding locking right now. I think it's fine to increase the error margin so we get predictable results